### ISSUE TITLE
Allow structured generation through "response_format" in mlx_server

### ISSUE THREAD

**Issue by avbiswas** [ml-explore/mlx-lm#1007]
Currently passing `response_format` into client.chat.completions.create after starting a mlx_server returns an empty output (not even an error, which makes it harder to catch).

Can we integrate outlines (or something similar) and generate structured outputs? While still maintaining continuous batching and parallel decoding?

Outlines supports basic generate already.
https://dottxt-ai.github.io/outlines/latest/features/models/mlxlm/

**Comment by LeaveNhA** [comment]
Is this issue still available? Why we are not solving it? A priority problem?

**Comment by 000alen** [comment]
any update on this?

**Comment by katlun-lgtm** [comment]
I'd like to take a run at native structured output (`response_format`), but before writing a large PR I want to agree on the one decision that will make or break it: **which grammar-matcher backend you'd accept.**

### Current state (so we're on the same page)

The enabling hook already exists — a constraint is just a logits processor. `generate_step` runs, per token:

```python
# mlx_lm/generate.py (~L415)
for processor in logits_processors:      # Callable[[mx.array, mx.array], mx.array]
    logits = processor(tokens, logits)
```

So a schema/grammar constraint is a processor that, given `tokens` so far, masks disallowed next-token ids to `-inf`. What's missing is (a) the matcher that produces the allowed-token set incrementally, (b) assembly in `make_logits_processors` (`sample_utils.py`), and (c) `server.py` reading `response_format` (it currently ignores it, which is why the request silently returns unconstrained text). There's no grammar engine in the tree today, so this is greenfield behind an existing, stable hook.

### Proposed shape

- A `StructuredLogitsProcessor` wrapping an **incremental matcher**: after each accepted token it yields the set of valid next token ids; the processor masks the rest.
- **Phase 1: JSON + JSON-Schema** (`response_format={"type": "json_object"}` and `{"type": "json_schema", "json_schema": {...}}`) — covers the large majority of agent/tool use.
- Wire `response_format` in `server.py` → build the processor → existing generate path. Also expose via `generate()` / CLI.
- **Phase 2 (follow-ups): regex/CFG.** Batching looks largely reachable already — `BatchGenerator` threads per-sequence `logits_processors` (generate.py ~L1069), so a per-sequence grammar processor should extend there with little new machinery (relates to #852).

### The decision I'd like your call on — matcher backend

1. **Vendor/depend on a proven engine** — `xgrammar` (vLLM/SGLang), `llguidance`, or `outlines-core`. Best correctness + speed, but adds a compiled/Rust dependency.
2. **A slim internal pure-Python JSON-Schema matcher** — zero new deps, keeps mlx-lm light, JSON-schema-only to start, and I carry the correctness burden.

mlx-lm has stayed dependency-light, so I lean toward **(2)** for a first PR — a no-dep JSON-Schema FSM behind an interface that could later swap to (1) for full CFG — but I don't want to send an unsolicited ~800-line PR in the wrong shape. Your call on the backend decides everything downstream.

### Two implementation notes

- **Perf:** the per-step allowed-token mask over the full vocab is the hot path. I'd keep the matcher's next-token-set computation off the critical path and do the mask-apply as an `mx` op (or a small Metal kernel if it profiles as worth it). I have an M3 Max and am happy to own the kernel/benchmark side.
- **Correctness:** the subtle part is tokenizer alignment — byte-level BPE and multi-byte/partial-UTF-8 (CJK) tokens straddling grammar boundaries. I'd build against a byte-level test suite from the start.

### Questions

1. **Backend:** no-dep internal JSON-Schema FSM, or a vendored engine (`xgrammar` / `llguidance` / `outlines-core`)?
2. **Surface:** is `response_format` on the server the right entry point, matching OpenAI's `{"type": "json_schema", "json_schema": {...}}`?
3. **Scope for PR #1:** JSON-object + JSON-Schema, single-sequence, with batch (#852) + regex/CFG as follow-ups — acceptable?

Happy to start as soon as you point me at the shape you'd merge.

**Comment by leee-dw** [comment]
A data point from running this in production: I wired Outlines into `mlx_lm.server` as an out-of-tree wrapper (no changes to installed files) and have been serving constrained JSON from it for a personal automation stack. It works, and the hook surface @katlun-lgtm described is indeed enough — but **the batching interaction is a real trap**, which is exactly what the original question asked about.

mlx-lm 0.31.3, `mlx-community/gemma-4-26b-a4b-it-6bit`.

### What the wrapper touches

Three seams, all monkey-patched so an upgrade can't silently keep a stale patch:

1. `APIHandler.handle_completion` — pull `response_format` off the request body.
2. `GenerationArguments` — carry the schema on the args object. This one matters: generation runs on a separate thread (`Thread(target=self._generate)`), so thread-locals don't survive the handoff; the args object does.
3. `_make_logits_processors` — append the Outlines processor.

```python
from outlines.backends import get_json_schema_logits_processor
from outlines.models.mlxlm import from_mlxlm

wrapper = from_mlxlm(provider.model, provider.tokenizer)   # cache per model
processor = get_json_schema_logits_processor(None, wrapper, schema)  # per request
```

Outlines already normalizes 1-D `input_ids` to 2-D specifically for mlx-lm, so the processor drops straight into `logits_processors` with no adapter.

Two things worth knowing if someone writes the native version:

- The processor is **stateful** (it advances an FSM per token), so it has to be built per request. Building one costs ~0.27 s for a moderately nested schema — under 1% of a multi-second request, so caching it isn't worth the aliasing risk.
- Top-level array schemas work fine here, which is useful because the OpenAI `json_schema` spec only allows objects. Several of my call sites return arrays.

### The batching trap

This is the part I'd flag for any native implementation. With the server's batching path, **the constraint quietly stops applying after the first request.**

Observed, same prompt, same schema, twice in a row:

| | processor invocations | output |
|---|---|---|
| first request (single path) | 23 | valid, schema-conforming |
| second request (batch path) | 1 | well-formed JSON that **violates the schema** |

The failure is nasty because the response still looks like JSON — only the schema is wrong — so a client that just calls `json.loads()` won't notice.

I want to be careful about the cause: I did **not** pin it down. Reading 0.31.3, the batch path looks like it's meant to handle processors — `PromptProcessingBatch.generate` passes them into `GenerationBatch`, `extend` merges them, and `GenerationBatch._step` calls them under `if any(self.logits_processors)`. So this isn't simply "not implemented"; something subtler is dropping them after the first step. Someone who knows that code will likely spot it much faster than I would.

My workaround is to avoid the batch path entirely for constrained requests: `ResponseGenerator._is_batchable` returns `self.model_provider.is_batchable and args.seed is None`, so filling in a random seed routes them through `stream_generate`. That's fine for me (the server runs with `--prompt-concurrency 1 --decode-concurrency 1` anyway), but it's obviously not an answer for anyone who wants constrained output *and* continuous batching — which is what this issue is actually asking for.

Since silent degradation is the worst outcome here, the wrapper counts processor invocations per request and logs an error when a constrained request finishes with zero. That check is what surfaced the problem in the first place; a native implementation may want something similar under a debug flag.

### Performance

Contrary to the usual assumption that constraints cost speed, constrained generation came out **faster** — the model stops emitting preamble, code fences and filler:

| | parses cleanly | mean |
|---|---|---|
| prompt-only "return JSON" | 12/12 | 13.1 s |
| schema-constrained | 12/12 | 10.5 s |

(12 prompt variants of a classification task, temperature 0, thinking off.)

One caveat for thinking models: the constraint applies from the very first token, so it has to be paired with thinking disabled — otherwise the reasoning block is masked out the moment it starts. My wrapper warns instead of constraining when it sees that combination.

Happy to share the full wrapper if it's useful, though for the native version I suspect the batch-path question above is the more valuable part.
